src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdater.cs

```diff
@@ -122,6 +130,50 @@ private static void CancelAndDispose(CancellationTokenSource cancellationSource)
                 // dispose cancellation token source
                 cancellationSource.Dispose();
             }
+
+            private void PushTextChanges(Document oldDocument, Document newDocument)
+            {
+                SourceText oldText = null;
+                SourceText newText = null;
+                if ((oldDocument?.TryGetText(out oldText) == false) ||
+                    (newDocument?.TryGetText(out newText) == false))
+                {
+                    // we only support case where text already exist
+                    return;
+                }
+
+                // get text changes
+                var textChanges = newText.GetTextChanges(oldText);
+                if (textChanges.Count == 0)
+                {
+                    // no changes
+                    return;
+                }
+
+                // whole document case
+                if (textChanges.Count == 1 && textChanges[0].Span.Length == oldText.Length)
+                {
+                    // no benefit here. pulling from remote host is more efficient
+                    return;
+                }
+
+                // only cancelled when remote host gets shutdown
+                var token = Listener.BeginAsyncOperation(nameof(PushTextChanges));
+                _textChangeQueue.ScheduleTask(async () =>
```

Comment:
are there any race conditions we have to be worried about.  For example, could we hear about two changes to a doc in an overlapping manner in the OnWorkspaceChanged callback?  If so, could we end up enqueuing a later text change without a previous one?

Reply:
everything is done stateless. it send out base text's checksum and change over it. so even ordering doesn't matter. we keep the ordering since that's more efficient since checksum is most likely there for every one involved.

Reply:
I like that.  It's a very clean model.  But we should doc this well to make that clear.